### ISSUE TITLE
Training with multi-GPUs cannot converge

### ISSUE THREAD

**Issue by deargqy** [tensorpack/benchmarks#8]
Hello,

I use the same(only modify the ipc address from 'ipc://@imagenet-train-b{}'.format(batch) to 'ipc:///tmp' ) py scripts from benchmarks/ResNet-Horovod to train the baseline model. But unfortunately, I found the result did not converge as expected. The environment and commands I used are listed below:

Python version: python3
Hhorovod version:0.16.0
Tensorpack version: 0.9.4

mpirun --allow-run-as-root -np 4 python3 imagenet-resnet-horovod.py  -d 50 --logdir ${LOG_DIR} --data /data/glusterfs_data/11082824/Data/CLS-LOC/  --batch 64 --no-zmq-ops

and  some of the results are:
[0416 05:11:23 @monitor.py:467] train-error-top1: 0.64265
[0416 05:24:37 @monitor.py:467] train-error-top1: 0.63513
[0416 05:36:53 @monitor.py:467] train-error-top1: 0.63307
[0416 05:48:24 @monitor.py:467] train-error-top1: 0.63591
[0416 06:00:05 @monitor.py:467] train-error-top1: 0.64311
[0416 06:12:09 @monitor.py:467] train-error-top1: 0.62931
[0416 06:23:40 @monitor.py:467] train-error-top1: 0.62378
[0416 06:34:44 @monitor.py:467] train-error-top1: 0.62418
[0416 06:45:55 @monitor.py:467] train-error-top1: 0.6367
[0416 06:56:54 @monitor.py:467] train-error-top1: 0.62604
[0416 07:07:54 @monitor.py:467] train-error-top1: 0.6293
[0416 07:19:04 @monitor.py:467] train-error-top1: 0.62163
[0416 07:30:15 @monitor.py:467] train-error-top1: 0.61299
[0416 07:41:18 @monitor.py:467] train-error-top1: 0.6069
[0416 07:53:04 @monitor.py:467] train-error-top1: 0.60632
[0416 08:04:33 @monitor.py:467] train-error-top1: 0.60605
[0416 08:16:11 @monitor.py:467] train-error-top1: 0.61349
[0416 08:28:04 @monitor.py:467] train-error-top1: 0.60073
[0416 08:40:10 @monitor.py:467] train-error-top1: 0.60852
[0416 08:52:47 @monitor.py:467] train-error-top1: 0.59606
[0416 09:05:11 @monitor.py:467] train-error-top1: 0.59727
[0416 09:17:37 @monitor.py:467] train-error-top1: 0.60537
[0416 09:30:35 @monitor.py:467] train-error-top1: 0.60737
[0416 09:42:55 @monitor.py:467] train-error-top1: 0.59626
[0416 09:55:24 @monitor.py:467] train-error-top1: 0.61557
[0416 10:08:30 @monitor.py:467] train-error-top1: 0.60756
[0416 10:26:31 @monitor.py:467] train-error-top1: 0.58888
[0416 10:41:43 @monitor.py:467] train-error-top1: 0.61388
[0416 10:54:00 @monitor.py:467] train-error-top1: 0.62244
[0416 11:06:38 @monitor.py:467] train-error-top1: 0.59488
[0416 11:18:57 @monitor.py:467] train-error-top1: 0.60776
[0416 11:31:28 @monitor.py:467] train-error-top1: 0.60432
[0416 11:44:13 @monitor.py:467] train-error-top1: 0.59858

As you can see, even I finished running all the epochs the top1 error is still very high. 

Thanks for your help

**Comment by ppwwyyxx** [comment]
Please post your `git diff` and training logs.

**Comment by deargqy** [comment]
The git diff is :
--- a/benchmarks/ResNet-Horovod/imagenet-resnet-horovod.py
+++ b/imagenet-resnet-horovod.py
@@ -29,7 +29,7 @@ class Model(ImageNetModel):
         }[depth]
 
     def get_logits(self, image):
-        with argscope([Conv2D, MaxPooling, GlobalAvgPooling, BatchNorm], data_format='NCHW'):
+        with argscope([Conv2D, MaxPooling, GlobalAvgPooling, BatchNorm], data_format='NHWC'):
             return resnet_backbone(image, self.num_blocks, resnet_group, resnet_bottleneck)
 
 
@@ -62,9 +62,9 @@ def get_config(model, fake=False):
         steps_per_epoch = 50
     else:
         logger.info("#Tower: {}; Batch size per tower: {}".format(hvd.size(), batch))
         logger.info("#Tower: {}; Batch size per tower: {}".format(hvd.size(), batch))
         logger.info("#Tower: {}; Batch size per tower: {}".format(hvd.size(), batch))
-        zmq_addr = 'ipc://@imagenet-train-b{}'.format(batch)
+        zmq_addr = 'ipc:///tmp/zhiwennb'
         if args.no_zmq_ops:
-            dataflow = RemoteDataZMQ(zmq_addr, hwm=150, bind=False)
+            dataflow = RemoteDataZMQ(zmq_addr, hwm=750, bind=False)
             data = QueueInput(dataflow)
         else:
             data = ZMQInput(zmq_addr, 30, bind=False)
@@ -126,6 +126,7 @@ def get_config(model, fake=False):
         callbacks=callbacks,
         steps_per_epoch=steps_per_epoch,
         max_epoch=35 if args.fake else 95,
+        session_init=SaverRestore(args.load) if args.load is not None else None,
     )
 
 
@@ -149,7 +150,7 @@ if __name__ == '__main__':
     """
     parser.add_argument('--batch', help='per-GPU batch size', default=32, type=int)
     args = parser.parse_args()
-
+    os.environ["HOROVOD_STALL_CHECK_DISABLE"] = '1'
     if args.eval:
         batch = 128    # something that can run on one gpu
         ds = get_val_dataflow(args.data, batch, fbresnet_augmentor(False))
@@ -160,7 +161,7 @@ if __name__ == '__main__':
     logger.info("Training on {}".format(socket.gethostname()))
     # Print some information for sanity check.
     os.system("nvidia-smi")
-    assert args.load is None
+    #assert args.load is None
 
     hvd.init()

The logs of last two epoch are shown below. If you need all of them, I can send to you.
[32m[0416 11:18:55 @base.py:275][0m Start Epoch 94 ...
  0%|          |0/5005[00:00<?,?it/s]
[32m[0416 11:18:55 @base.py:275][0m Start Epoch 94 ...
  0%|          |0/5005[00:00<?,?it/s]
[32m[0416 11:18:57 @saver.py:79][0m Model saved to /data/glusterfs_data/11082824/models/logs_1pod/model-465465.
[32m[0416 11:18:57 @misc.py:109][0m Estimated Time Left: 28 minutes 10 seconds
[32m[0416 11:18:57 @monitor.py:467][0m QueueInput/queue_size: 1.1755e-38
[32m[0416 11:18:57 @monitor.py:467][0m l2_regularize_loss: 0.45162
[32m[0416 11:18:57 @monitor.py:467][0m learning_rate: 0.0001
[32m[0416 11:18:57 @monitor.py:467][0m train-error-top1: 0.60776
[32m[0416 11:18:57 @monitor.py:467][0m train-error-top5: 0.38325
[32m[0416 11:18:57 @monitor.py:467][0m xentropy-loss: 2.9718
[32m[0416 11:18:57 @base.py:275][0m Start Epoch 94 ...
  0%|          |0/5005[00:00<?,?it/s]
[32m[0416 11:31:26 @base.py:285][0m Epoch 94 (global_step 470470) finished, time:12 minutes 31 seconds.
[32m[0416 11:31:26 @trainers.py:422][0m Running horovod broadcast ...
[32m[0416 11:31:26 @base.py:285][0m Epoch 94 (global_step 470470) finished, time:12 minutes 31 seconds.
[32m[0416 11:31:26 @trainers.py:422][0m Running horovod broadcast ...
[32m[0416 11:31:26 @base.py:285][0m Epoch 94 (global_step 470470) finished, time:12 minutes 31 seconds.
[32m[0416 11:31:26 @trainers.py:422][0m Running horovod broadcast ...
[32m[0416 11:31:26 @base.py:285][0m Epoch 94 (global_step 470470) finished, time:12 minutes 29 seconds.
[32m[0416 11:31:26 @trainers.py:422][0m Running horovod broadcast ...
[32m[0416 11:31:26 @base.py:275][0m Start Epoch 95 ...

[32m[0416 11:31:26 @base.py:275][0m Start Epoch 95 ...
  0%|          |0/5005[00:00<?,?it/s]
[32m[0416 11:31:26 @base.py:275][0m Start Epoch 95 ...
  0%|          |0/5005[00:00<?,?it/s]
[32m[0416 11:31:28 @saver.py:79][0m Model saved to /data/glusterfs_data/11082824/models/logs_1pod/model-470470.
[32m[0416 11:31:28 @misc.py:109][0m Estimated Time Left: 12 minutes 59 seconds
[32m[0416 11:31:28 @monitor.py:467][0m QueueInput/queue_size: 0.0625
[32m[0416 11:31:28 @monitor.py:467][0m l2_regularize_loss: 0.4513
[32m[0416 11:31:28 @monitor.py:467][0m learning_rate: 0.0001
[32m[0416 11:31:28 @monitor.py:467][0m train-error-top1: 0.60432
[32m[0416 11:31:28 @monitor.py:467][0m train-error-top5: 0.38028
[32m[0416 11:31:28 @monitor.py:467][0m xentropy-loss: 2.9774
[32m[0416 11:31:28 @base.py:275][0m Start Epoch 95 ...
  0%|          |0/5005[00:00<?,?it/s]
[32m[0416 11:44:10 @base.py:285][0m Epoch 95 (global_step 475475) finished, time:12 minutes 43 seconds.
[32m[0416 11:44:10 @trainers.py:422][0m Running horovod broadcast ...
[32m[0416 11:44:10 @base.py:285][0m Epoch 95 (global_step 475475) finished, time:12 minutes 43 seconds.
[32m[0416 11:44:10 @trainers.py:422][0m Running horovod broadcast ...
[32m[0416 11:44:10 @base.py:285][0m Epoch 95 (global_step 475475) finished, time:12 minutes 43 seconds.
[32m[0416 11:44:10 @trainers.py:422][0m Running horovod broadcast ...
[32m[0416 11:44:10 @base.py:285][0m Epoch 95 (global_step 475475) finished, time:12 minutes 41 seconds.
[32m[0416 11:44:10 @trainers.py:422][0m Running horovod broadcast ...
[32m[0416 11:44:10 @base.py:289][0m Training has finished!
[32m[0416 11:44:10 @base.py:289][0m Training has finished!
[32m[0416 11:44:10 @base.py:289][0m Training has finished!
[32m[0416 11:44:10 @input_source.py:178][0m EnqueueThread QueueInput/input_queue Exited.


[32m[0416 11:44:13 @saver.py:79][0m Model saved to /data/glusterfs_data/11082824/models/logs_1pod/model-475475.
[32m[0416 11:44:13 @monitor.py:467][0m QueueInput/queue_size: 3.9482e-31
[32m[0416 11:44:13 @monitor.py:467][0m l2_regularize_loss: 0.45098
[32m[0416 11:44:13 @monitor.py:467][0m learning_rate: 0.0001
[32m[0416 11:44:13 @monitor.py:467][0m train-error-top1: 0.59858
[32m[0416 11:44:13 @monitor.py:467][0m train-error-top5: 0.38767
[32m[0416 11:44:13 @monitor.py:467][0m xentropy-loss: 2.9617
[32m[0416 11:44:13 @group.py:48][0m Callbacks took 3.097 sec in total. ModelSaver: 2.5 seconds
[32m[0416 11:44:13 @base.py:289][0m Training has finished!
[32m[0416 11:44:13 @input_source.py:178][0m EnqueueThread QueueInput/input_queue Exited.

Thanks

**Comment by ppwwyyxx** [comment]
You modification on the architecture is wrong. Please don't
